### ISSUE TITLE
[BUG] `flatMap` doesn't allocation space for registers except for the first time it executes

### ISSUE THREAD

**Issue by j-mie6** [j-mie6/parsley#166]
## Description
When a `flatMap` executes, it should allocate enough space for the registers inside it to function. But instead it only does this the first time and crashes the other times.

## Reproduction Steps
The following parser crashes, because the register array is empty on the second parse of `p`. 

```scala
import parsley.registers._
val r = Reg.make[Int]
val p = pure(7).flatMap { _ =>
    r.put(4) *> r.get
}
(p *> p).parse("")
```

## Expected behavior
The registers should be accounted for on all runs. The second run does not generate a `CalleeSave` instruction

## Additional context
**Parsley Version:** 4.2.6
**Scala Version:** 2.13.10

The issue here is that the decision to allocate registers to a `DynCall` is done by asking if there are any unallocated registers in scope. This is insufficient, because they will be allocated next round. Instead we need to find all the registers inside the subclass that the parent does not also allocate. Good luck.

**Comment by j-mie6** [comment]
Fixed in 4.2.7

**Comment by j-mie6** [comment]
Backport required on the 4.1 series.
